plutus-ledger/src/Ledger/Tx.hs

```diff
@@ -75,99 +73,119 @@ import Data.Proxy (Proxy (Proxy))
 import Data.Set (Set)
 import Data.Set qualified as Set
 import GHC.Generics (Generic)
-import Ledger.Address (Address, PaymentPubKey, StakePubKey, pubKeyAddress)
+import Ledger.Address (PaymentPubKey, StakePubKey, pubKeyAddress)
 import Ledger.Crypto (Passphrase, signTx, signTx', toPublicKey)
 import Ledger.Orphans ()
 import Ledger.Slot (SlotRange)
 import Ledger.Tx.CardanoAPI (SomeCardanoApiTx (SomeTx), ToCardanoError (..))
 import Ledger.Tx.CardanoAPI qualified as CardanoAPI
 import Ledger.Tx.Internal hiding (updateUtxoCollateral)
-import Plutus.Script.Utils.V1.Scripts (datumHash)
+import Plutus.Script.Utils.V1.Scripts (datumHash, validatorHash)
 import Plutus.V1.Ledger.Api qualified as V1
-import Plutus.V2.Ledger.Api qualified as V2
 -- for re-export
+import Plutus.V1.Ledger.Address qualified as V1
 import Plutus.V1.Ledger.Tx
 import Plutus.V1.Ledger.Tx qualified as V1.Tx
 import Prettyprinter (Pretty (pretty), braces, colon, hang, nest, viaShow, vsep, (<+>))
 
 type PrivateKey = Crypto.XPrv
 
--- | Transaction output that comes from a chain index query.
---
--- It is defined here instead of the plutus-chain-index because plutus-ledger
--- uses that datatype, and plutus-ledger can't depend on plutus-chain-index
--- because of a cyclic dependency.
---
--- This datatype was created in order to be used in
--- 'Ledger.Constraints.processConstraint', specifically with the constraints
--- 'MustSpendPubKeyOutput' and 'MustSpendScriptOutput'.
-data ChainIndexTxOut =
-    PublicKeyChainIndexTxOut {
-      -- | Address of the transaction output. The address is protected by a
-      -- public key hash.
-      _ciTxOutAddress         :: Address,
+-- | Transaction output represented off-chain
+-- add more
+data OffChainTxOut =
```

Comment:
Curious as to why you went from `DecoratedTxOut` to `OffChainTxOut`?

Reply:
I got some feedbeck that
1) having two constructors is useful
2) given this type represents a txOut as seen off-chain `OffChainTxOut` is a good name. `TxOut` is the on-chain version.
3) someone suggested "isn't just renaming enough", and this version is a simpler approch, mostly a rename with some tightening of the types.

Basically `OffChainTxOut` is made by
1) embedding the two constructors of `Credential` into the type, to retain evidence of the type of the address.
2) Turning `Either V1.ValidatorHash V1.Validator` -> `Maybe V1.Validator` since we already have the validator hash thanks to the above
3) Split `Either V1.DatumHash V1.Datum` into `Maybe DatumHash`, `Maybe Datum` to keep them both so we don't have to recompute the hash.
4) Expand `V2.OutputDatum` so we don't have to depend on V2, since we actually don't need to.